R/pmcmc.R

```diff
@@ -66,6 +65,59 @@ pmcmc <- function(pars, filter, initial = NULL, control = NULL) {
 }
 
 
+##' Run a pMCMC, with sensible random number behaviour, but schedule
+##' execution of the chains yourself. Use this if you want to
+##' distribute chains over (say) the nodes of an HPC system.
+##'
+##' @title pMCMC with manual chain scheduling
+##'
+##' @inheritParams pmcmc
+##' @export
+pmcmc_chains_prepare <- function(pars, filter, initial = NULL, control = NULL) {
+  assert_is(pars, c("pmcmc_parameters", "pmcmc_parameters_nested"))
+  assert_is(filter, "particle_filter")
+  assert_is(control, "pmcmc_control")
+
+  if (control$n_workers != 1) {
+    stop("'n_workers' must be 1")
+  }
+  if (!control$use_parallel_seed) {
+    stop("'use_parallel_seed' must be TRUE")
+  }
+
+  if (inherits(pars, "pmcmc_parameters_nested")) {
+    initial <- pmcmc_check_initial_nested(initial, pars, control$n_chains)
+  } else {
+    initial <- pmcmc_check_initial(initial, pars, control$n_chains)
+  }
+
+  seed <- make_seeds(control$n_chains, filter$inputs()$seed)
+
+  ret <- list(pars = pars, initial = initial, filter = filter,
+              control = control, seed = seed)
+  class(ret) <- "pmcmc_inputs"
+  ret
+}
+
+
+##' @param chain_id The chain index to run (1, 2, ..., `control$n_chains`)
+##'
+##' @param inputs A `pmcmc_inputs` object created by `pmcmc_chains_prepare`
+##'
+##' @export
+##' @rdname pmcmc_chains_prepare
+pmcmc_chains_run <- function(chain_id, inputs) {
```

Comment:
Very minor and subjective point but I personally find it easier to read:
```R
pmcmc_chains_run(inputs, 1)
pmcmc_chains_run(inputs, 2)
```

than

```R
pmcmc_chains_run(1, inputs)
pmcmc_chains_run(2, inputs)
```

Reply:
lapply disagrees with that, the current order is optimised to take advantage of that:

```
lapply(seq_len(nchains), pmcmc_chains_run, inputs)
```

the other way around you'd need to do

```
lapply(seq_len(nchains), pmcmc_chains_run, inputs = inputs) # or
lapply(seq_len(nchains), function(i) pmcmc_chains_run(inputs, i))
```